### PR TITLE
Fix IPv6 address format issue

### LINKED ISSUE #1906: swarm join with IPv6 not supported?
Hi,

It seems I am not able to to set the advertise option to the docker host's IPv6 address:

```
docker run --name swarm_agent -d swarm join --advertise=[2001:db8:85a3::8a2e:370:7334]:2375 consul://[2001:db8:85a3:d0c:1::2]:8500
docker logs swarm_agent
time="2016-02-29T13:22:15Z" level=fatal msg="--advertise should be of the form ip:port or hostname:port"
```

However, the discovery backend's URI is accepted if I use an IPv6 address.

### PR DESCRIPTION
**Pull request by dongluochen**
Fix #1906. In `github.com/docker/docker/pkg/discovery`, fix address formate issue with `net.JoinHostPort`. In `cli/join.go`, use `net.SplitHostPort` to parse address. 

```
dchen@vm3:~$ ./swarm join --addr=[2001:db8:0:f101::2]:4444 consul://192.168.56.202:8500/swarm
INFO[0000] Initializing discovery without TLS           
INFO[0000] Registering on the discovery service every 1m0s...  addr=[2001:db8:0:f101::2]:4444 discovery=consul://192.168.56.202:8500/swarm
```

```
dchen@vm3:~$ docker -H [2001:db8:0:f101::3]:4000 info
Containers: 18
 Running: 0
 Paused: 0
 Stopped: 18
Images: 24
Server Version: swarm/1.1.3
Role: primary
Strategy: spread
Filters: health, port, dependency, affinity, constraint
Nodes: 1
 vm3: [2001:db8:0:f101::2]:4444
  └ Status: Healthy
  └ Containers: 18
  └ Reserved CPUs: 0 / 2
  └ Reserved Memory: 0 B / 4.053 GiB
  └ Labels: executiondriver=native-0.2, kernelversion=3.19.0-25-generic, operatingsystem=Ubuntu 14.04.3 LTS, storagedriver=aufs
  └ Error: (none)
  └ UpdatedAt: 2016-03-02T17:56:12Z
Plugins: 
 Volume: 
 Network: 
Kernel Version: 3.19.0-25-generic
Operating System: linux
Architecture: amd64
CPUs: 2
Total Memory: 4.053 GiB
Name: vm2
```

### CODE REVIEW

File: cli/join.go [hunk 1]
```diff
@@ -11,7 +12,12 @@ import (
 )
 
 func checkAddrFormat(addr string) bool {
-	m, _ := regexp.MatchString("^[0-9a-zA-Z._-]+:[0-9]{1,5}$", addr)
+	// validate addr is in host:port form. Use net function to handle both IPv4/IPv6 cases.
+	_, port, err := net.SplitHostPort(addr)
+	if err != nil {
+		return false
+	}
+	m, _ := regexp.MatchString("^[0-9]{1,5}$", port)
```
Comment (aaronlehmann) [inline cli/join.go]: This would accept 99999 as a valid port. I think it's better to do `Atoi` and check that the result is within range.

Comment (dongluochen) [inline cli/join.go]: Agree.

**Comment by abronan** [comment]
LGTM

**Comment by nishanttotla** [comment]
LGTM

**Pull request merged**